File: mobile/packages/footprint-expo/src/utils/logger.ts
import * as Linking from 'expo-linking';

import sendSdkTelemetry from './send-sdk-telemetry';

const debugMode = false; // Enable this for local development

export const getStringMessage = (msg?: unknown | Error): string => {
  if (typeof msg === 'string') {
    return msg;
  }
  if (typeof msg !== 'object') {
    return 'Something went wrong';
  }
  if ((msg as Error).message) {
    return (msg as Error).message;
  }

  try {
    return JSON.stringify(msg);
  } catch {
    // Do nothing
  }
  return 'Something went wrong';
};

const messagePrefix = '@onefootprint/footprint-expo';

export const logError = (error: unknown) => {
  const errorMessage = getStringMessage(error);
  const errorMessageWithPrefix = `${messagePrefix}: ${errorMessage}`;
  if (debugMode) {
    console.error(errorMessageWithPrefix);
  } else {
    Linking.getInitialURL()
      .then(tenantDomain => {
        sendSdkTelemetry(errorMessage, 'error', tenantDomain ?? '');
      })
      .catch(() => {
        // Ignore any errors and send without tenant domain
        sendSdkTelemetry(errorMessage, 'error');
      });
  }

  return errorMessageWithPrefix;
};

export const logWarn = (error: unknown) => {
  const warnMessage = getStringMessage(error);
  const warnMessageWithPrefix = `${messagePrefix}: ${warnMessage}`;
  if (debugMode) {
    console.warn(warnMessageWithPrefix);
  } else {
    Linking.getInitialURL()
      .then(tenantDomain => {
        sendSdkTelemetry(warnMessage, 'warn', tenantDomain ?? '');
      })
      .catch(() => {
        // Ignore any errors and send without tenant domain
        sendSdkTelemetry(warnMessage, 'warn');
      });
  }

  return warnMessageWithPrefix;
};

File: mobile/packages/footprint_flutter/lib/src/utils/logger.dart
import 'dart:convert';

import 'package:footprint_flutter/src/config/constants.dart';
import 'package:footprint_flutter/src/utils/send_sdk_telemetry.dart';

enum LogLevel {
  error,
  warn,
}

String _getStringMessage(dynamic msg) {
  if (msg is String) {
    return msg;
  }
  if (msg is! Object) {
    return 'Something went wrong';
  }
  if (msg is Error) {
    return msg.toString();
  }
  try {
    return jsonEncode(msg);
  } catch (e) {
    // Do nothing
  }
  return 'Something went wrong';
}

String _log(LogLevel level, dynamic message, String sdkKind) {
  final String formattedMessage = _getStringMessage(message);
  final String messageWithPrefix = '$logPrefix: $formattedMessage';

  if (debugMode) {
    final logPrefix = level == LogLevel.error ? 'ERROR' : 'WARNING';
    print('$logPrefix: $messageWithPrefix');
  } else {
    sendSdkTelemetry(formattedMessage, level.toString(), sdkKind: sdkKind);
  }

  return messageWithPrefix;
}

String logError(dynamic message, {required String sdkKind}) {
  return _log(LogLevel.error, message, sdkKind);
}

String logWarn(dynamic message, {required String sdkKind}) {
  return _log(
    LogLevel.warn,
    message,
    sdkKind,
  );
}
